temporal-sdk/src/main/java/io/temporal/internal/worker/EagerActivitySlotsReservation.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.ScheduleActivityTaskCommandAttributes;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueResponse;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedRequest;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedResponse;
+
+public class EagerActivitySlotsReservation implements AutoCloseable {
+  private final EagerActivityDispatcher eagerActivityDispatcher;
+  private int acquiredReservationCount = 0;
+
+  EagerActivitySlotsReservation(
+      EagerActivityDispatcher eagerActivityDispatcher,
+      RespondWorkflowTaskCompletedRequest.Builder mutableRequest) {
+    this.eagerActivityDispatcher = eagerActivityDispatcher;
+    applyToRequest(mutableRequest);
+  }
+
+  public void applyToRequest(RespondWorkflowTaskCompletedRequest.Builder mutableRequest) {
+    for (int i = 0; i < mutableRequest.getCommandsCount(); i++) {
+      Command command = mutableRequest.getCommands(i);
+      if (command.getCommandType() != CommandType.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK) continue;
+
+      ScheduleActivityTaskCommandAttributes commandAttributes =
+          command.getScheduleActivityTaskCommandAttributes();
+      if (!commandAttributes.getRequestEagerExecution()) continue;
+
+      if (this.eagerActivityDispatcher.tryReserveActivitySlot(commandAttributes)) {
+        this.acquiredReservationCount++;
+      } else {
+        mutableRequest.setCommands(
+            i,
+            command.toBuilder()
+                .setScheduleActivityTaskCommandAttributes(
+                    commandAttributes.toBuilder().setRequestEagerExecution(false)));
+      }
+    }
+  }
+
+  public void handleResponse(RespondWorkflowTaskCompletedResponse serverResponse) {
+    if (serverResponse.getActivityTasksCount() > this.acquiredReservationCount)
+      throw new IllegalStateException(
+          String.format(
+              "Unexpectedly received %d eager activities though we only requested %d",
+              serverResponse.getActivityTasksCount(), this.acquiredReservationCount));
+
+    if (this.acquiredReservationCount == 0) return;
+
+    for (PollActivityTaskQueueResponse act : serverResponse.getActivityTasksList())
+      this.eagerActivityDispatcher.dispatchActivity(act);
```

Comment:
What if this throws after dispatching, say, 5? Instead of decrementing the outstanding count here, you're just setting to 0 at the end so when `close()` is called, it will release more than it still has. If you don't expect anything in this function to ever throw, it may be worth noting that.

It might be clearer to rename `acquiredReservationCount` to `outstandingReservedSlots` or something, so that you can decrement as you move past them and have close be exception safe. One might expect that member to _always_ have an accurate count, not just after a successful call here.

Reply:
Yeah... I debated this with myself while writing this code. The idea of using a try-with-resource actually came from this concern and the difficulty of properly handling the number of outstanding slot reservations counter. (and btw, thanks for the name, its clearer)

I don't think there's actually a risk that `dispatchActivity` might ever fail in our code, but just for the sake of clarity, I'm refactoring this...